commit 98c470841f2027e8591b5790540861c9bf5eb611
Author: Josef Johansson <josef@oderland.se>
Date:   Sat Oct 16 21:32:46 2021 +0200

    PCI/MSI: Re-add checks for skip masking MSI-X on Xen PV
    
    'commit fcacdfbef5a1 ("PCI/MSI: Provide a new set of mask and unmask
    functions")' introduced new functions that were missing checks for
    pci_msi_ignore_mask that existed in 'commit 446a98b19fd6 ("PCI/MSI: Use
    new mask/unmask functions")'. This patch adds them back since it was
    causing locks in GPU drivers under Xen.
    
    Fixes: fcacdfbef5a1 ("PCI/MSI: Provide a new set of mask and unmask functions")
    Signed-off-by: Josef Johansson <josef@oderland.se>
    Suggested-by: Jason Andryuk <jandryuk@gmail.com>
    Link: https://lore.kernel.org/linux-pci/CAKf6xpvGyCKVHsvauP54=0j10fxis4XiiqBNWH+1cpkbtt_QJw@mail.gmail.com/

diff --git a/drivers/pci/msi.c b/drivers/pci/msi.c
index 0099a00af361..355b791e382f 100644
--- a/drivers/pci/msi.c
+++ b/drivers/pci/msi.c
@@ -148,6 +148,9 @@ static noinline void pci_msi_update_mask(struct msi_desc *desc, u32 clear, u32 s
 	raw_spinlock_t *lock = &desc->dev->msi_lock;
 	unsigned long flags;
 
+	if (pci_msi_ignore_mask || desc->msi_attrib.is_virtual)
+		return;
+
 	raw_spin_lock_irqsave(lock, flags);
 	desc->msi_mask &= ~clear;
 	desc->msi_mask |= set;
@@ -181,6 +184,9 @@ static void pci_msix_write_vector_ctrl(struct msi_desc *desc, u32 ctrl)
 {
 	void __iomem *desc_addr = pci_msix_desc_addr(desc);
 
+	if (pci_msi_ignore_mask || desc->msi_attrib.is_virtual)
+		return;
+
 	writel(ctrl, desc_addr + PCI_MSIX_ENTRY_VECTOR_CTRL);
 }
 
